File: src/scoring.ts
import type { Json, Verdict } from "./types.js";

export function number(value: unknown, fallback: number | null = 0): number | null {
  if (value === null || value === undefined || value === "") return fallback;
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : fallback;
}
const flag = (v: unknown): boolean | null =>
  v === true || [1, "1", "yes", "true"].includes(v as any)
    ? true
    : v === false || [0, "0", "no", "false"].includes(v as any)
      ? false
      : null;
const firstNumber = (...values: unknown[]): number | null => {
  for (const value of values) {
    const parsed = number(value, null);
    if (parsed !== null) return parsed;
  }
  return null;
};
const firstPositiveNumber = (...values: unknown[]): number | null => {
  let zero: number | null = null;
  for (const value of values) {
    const parsed = number(value, null);
    if (parsed === null) continue;
    if (parsed > 0) return parsed;
    if (parsed === 0) zero = 0;
  }
  return zero;
};

const holderAnalysisFields = (info: Json): unknown[] => {
  const stat = info.stat ?? {};
  return [
    stat.top_rat_trader_percentage,
    stat.top_bundler_trader_percentage,
    stat.top_entrapment_trader_percentage,
    stat.top_bot_degen_percentage,
    stat.bot_degen_rate,
    stat.fresh_wallet_rate,
    stat.top_10_holder_rate,
    stat.dev_team_hold_rate,
    stat.creator_hold_rate,
    stat.private_vault_hold_rate,
  ];
};

export function hasPopulatedHolderAnalysis(info: Json): boolean {
  return holderAnalysisFields(info).some((value) => {
    const parsed = number(value, null);
    return parsed !== null && parsed !== 0;
  });
}

export function screenTrackedBuyToken(info: Json, security: Json, pool: Json, cfg: Json): Verdict {
  const reasons: string[] = [],
    warnings: string[] = [];
  let failures = 0;
  const required = (ok: boolean, label: string) => {
    reasons.push(`${ok ? "PASS" : "FAIL"} ${label}`);
    if (!ok) failures++;
  };
  const maxRate = (value: number | null, limit: number, label: string) =>
    required(
      value !== null && value <= limit,
      value === null
        ? `${label} unavailable`
        : `${label} ${(value * 100).toFixed(1)}% <= ${(limit * 100).toFixed(1)}%`,
    );
  const price = number(info.price?.price, null),
    supply = firstNumber(info.circulating_supply, info.total_supply),
    marketCap =
      firstNumber(info.market_cap, info.price?.market_cap) ??
      (price !== null && supply !== null ? price * supply : null),
    liquidity = firstNumber(pool.liquidity, info.liquidity, info.pool?.liquidity),
    holderCount = firstNumber(info.holder_count, info.stat?.holder_count),
    top10 = firstPositiveNumber(
      security.top_10_holder_rate,
      info.stat?.top_10_holder_rate,
      info.dev?.top_10_holder_rate,
    ),
    minLiquidity = number(cfg.tracked_alert_min_liquidity_usd, 500) ?? 500,
    minRatio = number(cfg.tracked_alert_min_liquidity_to_market_cap_ratio, 0.01) ?? 0.01,
    maxRatio = number(cfg.tracked_alert_max_liquidity_to_market_cap_ratio, 10) ?? 10,
    minMarketCap = number(cfg.tracked_alert_min_market_cap_usd, 5000) ?? 5000,
    minHolders = number(cfg.tracked_alert_min_holders, 10) ?? 10;
  const honeypot = flag(security.is_honeypot ?? security.honeypot),
    blacklist = flag(security.is_blacklist ?? security.blacklist),
    openSource = flag(security.is_open_source ?? security.open_source),
    renounced = flag(security.is_renounced ?? security.owner_renounced ?? security.renounced),
    locked = flag(security.lock_summary?.is_locked),
    cannotSell = flag(security.can_not_sell);

  required(Boolean(String(info.symbol ?? "").trim()), "GMGN token record exists");
  if (cfg.require_honeypot_false) {
    required(honeypot === false, "honeypot check passed");
    required(blacklist === false, "blacklist check passed");
    required(cannotSell !== true, "sellability check passed");
  } else if (honeypot === true) required(false, "honeypot detected");
  if (cfg.require_open_source) required(openSource === true, "contract source verified");
  if (cfg.require_owner_renounced) required(renounced === true, "contract ownership renounced");
  if (cfg.require_liquidity_locked) required(locked === true, "liquidity lock confirmed");
  if (cfg.require_renounced_mint)
    required(flag(security.renounced_mint) === true, "mint authority renounced");
  if (cfg.require_renounced_freeze)
    required(flag(security.renounced_freeze_account) === true, "freeze authority renounced");
  required(
    liquidity !== null && liquidity >= minLiquidity,
    `tracked-alert liquidity $${Math.round(liquidity ?? 0).toLocaleString()} >= $${minLiquidity.toLocaleString()}`,
  );
  const liquidityRatio =
    liquidity !== null && marketCap !== null && marketCap > 0 ? liquidity / marketCap : null;
  required(
    liquidityRatio !== null && liquidityRatio >= minRatio,
    liquidityRatio === null
      ? "liquidity-to-market-cap ratio unavailable"
      : `liquidity-to-market-cap ratio ${(liquidityRatio * 100).toFixed(2)}% >= ${(minRatio * 100).toFixed(2)}%`,
  );
  required(
    liquidityRatio !== null && liquidityRatio <= maxRatio,
    liquidityRatio === null
      ? "liquidity-to-market-cap ratio unavailable"
      : `liquidity-to-market-cap ratio ${(liquidityRatio * 100).toFixed(2)}% <= ${(maxRatio * 100).toFixed(2)}%`,
  );
  required(
    marketCap !== null && marketCap >= minMarketCap,
    `tracked-alert market cap $${Math.round(marketCap ?? 0).toLocaleString()} >= $${minMarketCap.toLocaleString()}`,
  );
  required(
    holderCount !== null && holderCount >= minHolders,
    `tracked-alert holders ${Math.round(holderCount ?? 0)} >= ${minHolders}`,
  );
  required(
    hasPopulatedHolderAnalysis(info),
    "GMGN holder analysis is populated; an all-zero block is not accepted as safe",
  );
  required(top10 !== null && top10 > 0, "top-10 concentration is populated and above 0%");
  maxRate(
    top10,
    number(cfg.tracked_alert_max_top_10_holder_rate, cfg.max_top_10_holder_rate) ??
      cfg.max_top_10_holder_rate,
    "top-10 concentration",
  );
  if (cfg.require_honeypot_false) {
    maxRate(number(security.buy_tax, null), cfg.max_buy_tax ?? 0.05, "buy tax");
    maxRate(number(security.sell_tax, null), cfg.max_sell_tax ?? 0.05, "sell tax");
  }
  const passed = reasons.filter((reason) => reason.startsWith("PASS ")).length;
  return {
    passed: failures === 0,
    score: Math.round((1000 * passed) / Math.max(reasons.length + warnings.length, 1)) / 10,
    reasons,
    warnings,
  };
}

export function wilsonLowerBound(wins: number, total: number, z = 1.96): number {
  if (total <= 0) return 0;
  const p = wins / total;
  return (
    (p + (z * z) / (2 * total) - z * Math.sqrt((p * (1 - p) + (z * z) / (4 * total)) / total)) /
    (1 + (z * z) / total)
  );
}

export function scoreToken(
  info: Json,
  security: Json,
  pool: Json,
  cluster: Json,
  cfg: Json,
): Verdict {
  const reasons: string[] = [],
    warnings: string[] = [];
  let failures = 0;
  const required = (ok: boolean, label: string) => {
    reasons.push(`${ok ? "PASS" : "FAIL"} ${label}`);
    if (!ok) failures++;
  };
  const knownMax = (value: number | null, limit: number, label: string) => {
    if (value === null) {
      warnings.push(`Unknown critical field: ${label}`);
      failures++;
    } else
      required(
        value <= limit,
        `${label} ${(value * 100).toFixed(1)}% <= ${(limit * 100).toFixed(1)}%`,
      );
  };
  const price = number(info.price?.price, null),
    supply = number(info.circulating_supply, null);
  const marketCap = price !== null && supply !== null ? price * supply : null;
  const liquidity = number(pool.liquidity ?? info.liquidity, null);
  const stat = info.stat ?? {};
  const honeypot = flag(security.is_honeypot),
    blacklist = flag(security.is_blacklist ?? security.blacklist),
    openSource = flag(security.is_open_source ?? security.open_source);
  const renounced = flag(security.is_renounced ?? security.owner_renounced ?? security.renounced);
  const locked = flag(security.lock_summary?.is_locked);
  if (cfg.require_honeypot_false) {
    required(honeypot === false, "honeypot check passed");
    required(blacklist === false, "blacklist check passed");
    required(flag(security.can_not_sell) !== true, "sellability check passed");
  } else if (honeypot === true) required(false, "honeypot detected");
  if (cfg.require_open_source) required(openSource === true, "contract source verified");
  if (cfg.require_owner_renounced) required(renounced === true, "contract ownership renounced");
  if (cfg.require_liquidity_locked) required(locked === true, "liquidity lock confirmed");
  if (cfg.require_renounced_mint)
    required(flag(security.renounced_mint) === true, "mint authority renounced");
  if (cfg.require_renounced_freeze)
    required(flag(security.renounced_freeze_account) === true, "freeze authority renounced");
  required(
    liquidity !== null && liquidity >= cfg.min_liquidity_usd,
    `liquidity $${Math.round(liquidity ?? 0).toLocaleString()}`,
  );
  const liquidityRatio =
      liquidity !== null && marketCap !== null && marketCap > 0 ? liquidity / marketCap : null,
    minRatio = number(cfg.min_liquidity_to_market_cap_ratio, 0.05) ?? 0.05;
  required(
    liquidityRatio !== null && liquidityRatio >= minRatio,
    liquidityRatio === null
      ? "liquidity-to-market-cap ratio unavailable"
      : `liquidity-to-market-cap ratio ${(liquidityRatio * 100).toFixed(2)}% >= ${(minRatio * 100).toFixed(2)}%`,
  );
  required(
    marketCap !== null &&
      marketCap >= cfg.min_market_cap_usd &&
      marketCap <= cfg.max_market_cap_usd,
    `market cap $${Math.round(marketCap ?? 0).toLocaleString()}`,
  );
  required(
    (number(info.holder_count, 0) ?? 0) >= cfg.min_holders,
    `holders ${number(info.holder_count, 0)}`,
  );
  required(
    (number(info.price?.volume_5m, 0) ?? 0) >= cfg.min_volume_5m_usd,
    `5m volume $${Math.round(number(info.price?.volume_5m, 0) ?? 0).toLocaleString()}`,
  );
  required(
    hasPopulatedHolderAnalysis(info),
    "GMGN holder analysis is populated; an all-zero block is not accepted as safe",
  );
  const top10 = firstPositiveNumber(
    security.top_10_holder_rate,
    stat.top_10_holder_rate,
    info.dev?.top_10_holder_rate,
  );
  required(top10 !== null && top10 > 0, "top-10 concentration is populated and above 0%");
  knownMax(top10, cfg.max_top_10_holder_rate, "top-10 concentration");
  knownMax(
    number(security.dev_team_hold_rate ?? stat.dev_team_hold_rate, null),
    cfg.max_dev_team_hold_rate,
    "dev-team holding",
  );
  knownMax(
    number(security.bundler_trader_amount_rate ?? stat.top_bundler_trader_percentage, null),
    cfg.max_bundler_rate,
    "bundler activity",
  );
  knownMax(number(stat.bot_degen_rate, null), cfg.max_bot_rate, "bot activity");
  knownMax(
    number(security.rat_trader_amount_rate ?? stat.top_rat_trader_percentage, null),
    cfg.max_rat_trader_rate,
    "insider/rat activity",
  );
  knownMax(
    number(stat.top_entrapment_trader_percentage, null),
    cfg.max_entrapment_rate ?? 1,
    "entrapment activity",
  );
  if (cfg.require_honeypot_false) {
    knownMax(number(security.buy_tax, null), cfg.max_buy_tax ?? 0.05, "buy tax");
    knownMax(number(security.sell_tax, null), cfg.max_sell_tax ?? 0.05, "sell tax");
  }
  const entry = number(cluster.median_entry_price_usd, null);
  if (price === null || entry === null || entry <= 0) {
    warnings.push("Cannot measure price chase from cluster entry");
    failures++;
  } else
    required(
      price / entry - 1 <= (cluster.max_price_chase_ratio ?? 0.15),
      `price moved ${((price / entry - 1) * 100).toFixed(1)}% since median tracked entry`,
    );
  const passed = reasons.filter((r) => r.startsWith("PASS ")).length;
  return {
    passed: failures === 0,
    score: Math.round((1000 * passed) / Math.max(reasons.length + warnings.length, 1)) / 10,
    reasons,
    warnings,
  };
}

export function clusterEvents(events: Json[], cfg: Json): Json[] {
  const grouped = new Map<string, Json[]>();
  for (const event of events)
    if (event.side === "buy" && event.is_open_or_close === 0 && event.base_address)
      grouped.set(event.base_address, [...(grouped.get(event.base_address) ?? []), event]);
  const clusters: Json[] = [];
  for (const [address, rows] of grouped) {
    rows.sort((a, b) => Number(a.timestamp) - Number(b.timestamp));
    for (let start = 0; start < rows.length; start++) {
      const window = rows
        .slice(start)
        .filter((r) => Number(r.timestamp) - Number(rows[start]!.timestamp) <= cfg.window_seconds);
      const unique = new Map<string, Json>();
      for (const row of window) if (row.maker && !unique.has(row.maker)) unique.set(row.maker, row);
      const selected = [...unique.values()];
      if (selected.length < cfg.min_qualified_wallets) continue;
      const amounts = selected.map((r) => number(r.amount_usd, 0) ?? 0).sort((a, b) => a - b);
      const prices = selected
        .map((r) => number(r.price_usd, 0) ?? 0)
        .filter(Boolean)
        .sort((a, b) => a - b);
      const median = (a: number[]) =>
        a.length % 2 ? a[(a.length - 1) / 2]! : (a[a.length / 2 - 1]! + a[a.length / 2]!) / 2;
      clusters.push({
        address,
        symbol: selected[0]?.base_token?.symbol,
        wallets: [...unique.keys()].sort(),
        events: selected,
        wallet_count: selected.length,
        aggregate_buy_usd: amounts.reduce((a, b) => a + b, 0),
        median_buy_usd: median(amounts),
        median_entry_price_usd: prices.length ? median(prices) : null,
        first_timestamp: Math.min(...selected.map((r) => Number(r.timestamp))),
        last_timestamp: Math.max(...selected.map((r) => Number(r.timestamp))),
      });
      break;
    }
  }
  return clusters;
}

File: scoring.py
from __future__ import annotations

import math
import statistics
import time
from dataclasses import dataclass
from typing import Any, Iterable


@dataclass(frozen=True)
class Verdict:
    passed: bool
    score: float
    reasons: tuple[str, ...]
    warnings: tuple[str, ...] = ()


def number(value: Any, default: float | None = 0.0) -> float | None:
    if value in (None, ""):
        return default
    try:
        result = float(value)
    except (TypeError, ValueError):
        return default
    return result if math.isfinite(result) else default


def ratio(numerator: Any, denominator: Any) -> float:
    den = number(denominator, 0.0) or 0.0
    return (number(numerator, 0.0) or 0.0) / den if den > 0 else 0.0


def wilson_lower_bound(wins: int, total: int, z: float = 1.96) -> float:
    if total <= 0:
        return 0.0
    p = wins / total
    denominator = 1 + z * z / total
    centre = p + z * z / (2 * total)
    spread = z * math.sqrt((p * (1 - p) + z * z / (4 * total)) / total)
    return (centre - spread) / denominator


def _profit_row(payload: dict[str, Any], period: str) -> dict[str, Any]:
    row = payload.get(period, {})
    return row if isinstance(row, dict) else {}


def score_wallet(
    profits: dict[str, dict[str, Any]],
    stats: dict[str, Any],
    activity: Iterable[dict[str, Any]],
    cfg: dict[str, Any],
) -> Verdict:
    reasons: list[str] = []
    warnings: list[str] = []
    hard_failures = 0

    p7 = _profit_row(profits, "7d")
    p30 = _profit_row(profits, "30d")
    pall = _profit_row(profits, "all")
    rp7 = number(p7.get("realized_profit"), 0.0) or 0.0
    rp30 = number(p30.get("realized_profit"), 0.0) or 0.0
    rpall = number(pall.get("total_realized_profit", pall.get("realized_profit")), 0.0) or 0.0
    roi30 = ratio(rp30, p30.get("realized_profit_cost"))
    roiall = ratio(rpall, pall.get("total_realized_profit_cost", pall.get("realized_profit_cost")))
    trades30 = int(number(p30.get("buy"), 0) or 0) + int(number(p30.get("sell"), 0) or 0)

    pnl_stat = stats.get("pnl_stat") or {}
    token_count = int(number(pnl_stat.get("token_num"), 0) or 0)
    winrate = number(pnl_stat.get("winrate"), 0.0) or 0.0
    wins = int(round(winrate * token_count))
    winrate_lb = wilson_lower_bound(wins, token_count)

    checks = [
        (rp7 >= cfg["min_realized_profit_7d"], f"7d realized P&L ${rp7:,.0f}"),
        (rp30 >= cfg["min_realized_profit_30d"], f"30d realized P&L ${rp30:,.0f}"),
        (rpall >= cfg["min_realized_profit_all"], f"all-time realized P&L ${rpall:,.0f}"),
        (roi30 >= cfg["min_roi_30d"], f"30d realized ROI {roi30:.1%}"),
        (roiall >= cfg["min_roi_all"], f"all-time realized ROI {roiall:.1%}"),
        (trades30 >= cfg["min_trades_30d"], f"30d trades {trades30}"),
        (trades30 <= cfg["max_trades_30d"], f"30d trades {trades30} (spray ceiling)"),
        (token_count >= cfg["min_tokens_30d"], f"30d token sample {token_count}"),
        (token_count <= cfg["max_tokens_30d"], f"30d tokens {token_count} (spray ceiling)"),
        (winrate >= cfg["min_win_rate_30d"], f"30d win rate {winrate:.1%}"),
        (winrate_lb >= cfg["min_wilson_win_rate_30d"], f"95% Wilson win-rate floor {winrate_lb:.1%}"),
    ]
    for ok, label in checks:
        reasons.append(("PASS " if ok else "FAIL ") + label)
        hard_failures += int(not ok)

    activities = list(activity)
    buys = [a for a in activities if (a.get("event_type") or a.get("type")) == "buy"]
    unique_tokens = {((a.get("token") or {}).get("address")) for a in activities}
    unique_tokens.discard(None)
    buy_sizes = [number(a.get("cost_usd"), 0.0) or 0.0 for a in buys]
    median_buy = statistics.median(buy_sizes) if buy_sizes else 0.0
    if activities:
        spray_ok = len(unique_tokens) <= cfg["max_unique_tokens_in_100_actions"]
        size_ok = median_buy >= cfg["min_median_buy_usd"]
        reasons.append(("PASS " if spray_ok else "FAIL ") + f"{len(unique_tokens)} unique tokens in {len(activities)} sampled actions")
        reasons.append(("PASS " if size_ok else "FAIL ") + f"median sampled buy ${median_buy:,.0f}")
        hard_failures += int(not spray_ok) + int(not size_ok)
    else:
        warnings.append("No activity sample; bot/spray and position-size checks are unknown")
        hard_failures += 1

    # Score rewards repeatable, realized performance while shrinking small samples.
    score = 0.0
    score += min(max(rp30 / 10_000, 0), 1) * 20
    score += min(max(rpall / 100_000, 0), 1) * 10
    score += min(max(roi30 / 0.25, 0), 1) * 20
    score += min(max(roiall / 0.15, 0), 1) * 10
    score += min(max(winrate_lb / 0.60, 0), 1) * 25
    score += min(max(token_count / 100, 0), 1) * 10
    score += 5 if rp7 >= 0 else 0
    return Verdict(hard_failures == 0, round(score, 1), tuple(reasons), tuple(warnings))


def _flag(value: Any) -> bool | None:
    if value is True or value in (1, "1", "yes", "true"):
        return True
    if value is False or value in (0, "0", "no", "false"):
        return False
    return None


def score_token(
    info: dict[str, Any],
    security: dict[str, Any],
    pool: dict[str, Any],
    cluster: dict[str, Any],
    cfg: dict[str, Any],
) -> Verdict:
    reasons: list[str] = []
    warnings: list[str] = []
    failures = 0

    price = number((info.get("price") or {}).get("price"), None)
    supply = number(info.get("circulating_supply"), None)
    market_cap = price * supply if price is not None and supply is not None else None
    liquidity = number(pool.get("liquidity", info.get("liquidity")), None)
    holders = int(number(info.get("holder_count"), 0) or 0)
    stat = info.get("stat") or {}
    top10 = number(security.get("top_10_holder_rate", stat.get("top_10_holder_rate")), None)
    dev_hold = number(security.get("dev_team_hold_rate", stat.get("dev_team_hold_rate")), None)
    bundler = number(security.get("bundler_trader_amount_rate", stat.get("top_bundler_trader_percentage")), None)
    bot = number(stat.get("bot_degen_rate"), None)
    rat = number(security.get("rat_trader_amount_rate", stat.get("top_rat_trader_percentage")), None)
    entrapment = number(stat.get("top_entrapment_trader_percentage"), None)
    volume5m = number((info.get("price") or {}).get("volume_5m"), None)
    mint = _flag(security.get("renounced_mint"))
    freeze = _flag(security.get("renounced_freeze_account"))
    honeypot = _flag(security.get("is_honeypot"))
    open_source = _flag(security.get("is_open_source", security.get("open_source")))
    owner_renounced = _flag(security.get("is_renounced", security.get("owner_renounced", security.get("renounced"))))
    lock_summary = security.get("lock_summary") or {}
    liquidity_locked = _flag(lock_summary.get("is_locked"))
    buy_tax = number(security.get("buy_tax"), None)
    sell_tax = number(security.get("sell_tax"), None)

    def required(ok: bool, label: str) -> None:
        nonlocal failures
        reasons.append(("PASS " if ok else "FAIL ") + label)
        failures += int(not ok)

    def known_max(value: float | None, limit: float, label: str) -> None:
        nonlocal failures
        if value is None:
            warnings.append(f"Unknown critical field: {label}")
            failures += 1
        else:
            required(value <= limit, f"{label} {value:.1%} <= {limit:.1%}")

    if cfg.get("require_honeypot_false"):
        required(honeypot is False, "honeypot check passed")
    elif honeypot is True:
        required(False, "honeypot detected")
    if cfg.get("require_open_source"):
        required(open_source is True, "contract source verified")
    if cfg.get("require_owner_renounced"):
        required(owner_renounced is True, "contract ownership renounced")
    if cfg.get("require_liquidity_locked"):
        required(liquidity_locked is True, "liquidity lock confirmed")
    if cfg.get("require_renounced_mint"):
        required(mint is True, "mint authority renounced")
    if cfg.get("require_renounced_freeze"):
        required(freeze is True, "freeze authority renounced")
    required(liquidity is not None and liquidity >= cfg["min_liquidity_usd"], f"liquidity ${liquidity or 0:,.0f}")
    required(market_cap is not None and cfg["min_market_cap_usd"] <= market_cap <= cfg["max_market_cap_usd"], f"market cap ${market_cap or 0:,.0f}")
    required(holders >= cfg["min_holders"], f"holders {holders}")
    required(volume5m is not None and volume5m >= cfg["min_volume_5m_usd"], f"5m volume ${volume5m or 0:,.0f}")
    known_max(top10, cfg["max_top_10_holder_rate"], "top-10 concentration")
    known_max(dev_hold, cfg["max_dev_team_hold_rate"], "dev-team holding")
    known_max(bundler, cfg["max_bundler_rate"], "bundler activity")
    known_max(bot, cfg["max_bot_rate"], "bot activity")
    known_max(rat, cfg["max_rat_trader_rate"], "insider/rat activity")
    known_max(entrapment, cfg.get("max_entrapment_rate", 1.0), "entrapment activity")
    if cfg.get("require_honeypot_false"):
        known_max(buy_tax, cfg.get("max_buy_tax", 0.05), "buy tax")
        known_max(sell_tax, cfg.get("max_sell_tax", 0.05), "sell tax")

    cluster_price = number(cluster.get("median_entry_price_usd"), None)
    if price is None or cluster_price is None or cluster_price <= 0:
        warnings.append("Cannot measure price chase from cluster entry")
        failures += 1
    else:
        chase = price / cluster_price - 1
        required(chase <= cluster.get("max_price_chase_ratio", 0.15), f"price moved {chase:.1%} since median tracked entry")

    passed_count = sum(r.startswith("PASS ") for r in reasons)
    score = 100 * passed_count / max(len(reasons) + len(warnings), 1)
    return Verdict(failures == 0, round(score, 1), tuple(reasons), tuple(warnings))


def cluster_events(events: Iterable[dict[str, Any]], cfg: dict[str, Any]) -> list[dict[str, Any]]:
    grouped: dict[str, list[dict[str, Any]]] = {}
    for event in events:
        if event.get("side") != "buy" or event.get("is_open_or_close") != 0:
            continue
        address = event.get("base_address")
        if address:
            grouped.setdefault(address, []).append(event)

    clusters: list[dict[str, Any]] = []
    for address, rows in grouped.items():
        rows.sort(key=lambda row: int(row.get("timestamp") or 0))
        for start in range(len(rows)):
            window = [r for r in rows[start:] if int(r.get("timestamp") or 0) - int(rows[start].get("timestamp") or 0) <= cfg["window_seconds"]]
            by_wallet: dict[str, dict[str, Any]] = {}
            for row in window:
                maker = row.get("maker")
                if maker and maker not in by_wallet:
                    by_wallet[maker] = row
            selected = list(by_wallet.values())
            if len(selected) < cfg["min_qualified_wallets"]:
                continue
            amounts = [number(r.get("amount_usd"), 0.0) or 0.0 for r in selected]
            prices = [number(r.get("price_usd"), 0.0) or 0.0 for r in selected]
            prices = [p for p in prices if p > 0]
            clusters.append({
                "address": address,
                "symbol": ((selected[0].get("base_token") or {}).get("symbol")),
                "wallets": sorted(by_wallet),
                "events": selected,
                "wallet_count": len(selected),
                "aggregate_buy_usd": sum(amounts),
                "median_buy_usd": statistics.median(amounts),
                "median_entry_price_usd": statistics.median(prices) if prices else None,
                "first_timestamp": min(int(r.get("timestamp") or 0) for r in selected),
                "last_timestamp": max(int(r.get("timestamp") or 0) for r in selected),
            })
            break
    return clusters
